contrib/pg_tde/src/access/pg_tde_tdemap.c

```diff
@@ -1789,6 +1812,9 @@ pg_tde_put_key_into_cache(const RelFileLocator *rlocator, InternalKey *key)
 			elog(WARNING, "could not mlock internal key cache pages: %m");
 
 		tde_rel_key_cache.cap = (size - 1) / sizeof(RelKeyCacheRec);
+
+		/* update wal key pointers after moving the cache */
+		update_wal_keys_cache();
```

Comment:
Shouldn't this go to line 1808, after memcpy but before pfree?

Reply:
We don't care about old cache pages, moreover, we need the resized cache to be valid at this point.
update_wal_keys_cache() goes through the new (moved) cache and if it finds a WAL key, it takes its pointer and updates the fist member of the tde_wal_key_cache list, and advances. It doesn't care about previous pointers. We can safely assume that the order and amount of the keys remains the same after the resize.

Reply:
Yes, and it's singleprocess, should be okay.